vpr/src/analytical_place/flat_placement_mass_calculator.cpp

```diff
@@ -6,38 +6,182 @@
+        // Need to compute the number of bits for each of the ports.
+        int dual_port1_num_bits = 0;
+        if (num_address1_pins > 0) {
+            // FIXME: Found that the Titan architecture fails this test! Raise
+            //        an issue on this to maybe update the documentation or fix
+            //        the architecture.
+            //        See: prim_ram_1Kx9
+            //        Perhaps the architecture parsing file should check for this.
+            // VTR_ASSERT_MSG(num_address_pins == 0 && num_address2_pins > 0,
+            //                "Ill-formed dual port memory");
+            dual_port1_num_bits = (1 << num_address1_pins) * num_data_out1_pins;
+        }
+        int dual_port2_num_bits = 0;
+        if (num_address2_pins > 0) {
+            VTR_ASSERT_MSG(num_address_pins == 0 && num_address1_pins > 0,
+                           "Ill-formed dual port memory");
+            dual_port2_num_bits = (1 << num_address2_pins) * num_data_out2_pins;
+        }
+        // Note: We take the max of the two dual port num bits since dual ports
+        //       access the same memory but with different address bits.
+        int total_dual_port_num_bits = std::max(dual_port1_num_bits, dual_port2_num_bits);
+
+        // Get the total number of bits. We assume that this memory will be either
+        // a single port or dual port memory, therefore we can just sum these two
+        // terms. If it was a single port, the dual port num bits will be 0 and
+        // vice-versa.
+        int total_num_bits = single_port_num_bits + total_dual_port_num_bits;
+        VTR_ASSERT_MSG(total_num_bits != 0,
+                       "Cannot deduce number of bits in memory primitive.");
+        return total_num_bits;
+    }
+
+    // If this is not a memory, count the number of input and output pins.
+    // This makes physical blocks with more pins have more size than those with
+    // less. For example, a 4 lut would have a lower cost than a 5 lut.
+    float pb_cost = pb_type->num_input_pins + pb_type->num_output_pins + pb_type->num_clock_pins;
+    return pb_cost;
+}
+
+/**
+ * @brief Returns the mass of the given atom block.
+ */
+static float get_atom_mass(AtomBlockId blk_id, const Prepacker& prepacker, const AtomNetlist& atom_netlist) {
+
+    // Get the physical block that this block will most likely be packed into.
+    // This will give us a better estimate for how many resources this atom will
+    // use. For example, a LUT atom may only use 2 input pins; however, the
+    // architecture may only have 4-LUTs and 5-LUTs available. This method should
+    // return the smallest compatible resource for this atom (i.e. the 4-LUT).
+    const t_pb_graph_node* primitive = prepacker.get_expected_lowest_cost_pb_gnode(blk_id);
+
+    // Get the mass of this atom (i.e. how much resources on the device this atom
+    // will likely use).
+    float mass = 0.0f;
+    if (!is_primitive_memory_pb_type(primitive->pb_type)) {
```

Comment:
You might be able to shorten the code by calling your physical routine, and then just blending in two different proportions based on whether it is a memory or not.